h/oauth/grant_types/jwt_bearer.py

```diff
@@ -0,0 +1,53 @@
+# -*- coding: utf-8 -*-
+from oauthlib import oauth2
+from oauthlib.oauth2.rfc6749 import errors
+
+import logging
+log = logging.getLogger(__name__)
+
+
+class JWTBearerGrant(oauth2.ClientCredentialsGrant):
+    uri = 'urn:ietf:params:oauth:grant-type:jwt-bearer'
+
+    def validate_token_request(self, request):
+        if not getattr(request, 'grant_type'):
+            raise errors.InvalidRequestError('Request is missing grant type.',
+                                             request=request)
+
+        if not request.grant_type == self.uri:
+            raise errors.UnsupportedGrantTypeError(request=request)
+
+        if not getattr(request, 'assertion'):
+            raise errors.InvalidRequestError('Request is missing assertion.',
+                                             request=request)
+
+        for param in ('grant_type', 'scope'):
+            if param in request.duplicate_params:
+                raise errors.InvalidRequestError(
+                    'Duplicate %s parameter.' % param,
+                    request=request)
+
+        if request.client_id or request.client_assertion:
+            log.debug('Authenticating client, %r.', request)
+            if not self.request_validator.authenticate_client(request):
+                log.debug('Invalid client (%r), denying access.', request)
+                raise errors.InvalidClientError(request=request)
+            else:
+                request.client_id = request.client_id or \
+                    request.client.client_id
```

Comment:
Line 30 it may be `client_id` or `client_assertion` that triggers client authentication. After authentication, `request.client` should be set, so we ensure at that point we copy the `request.client.client_id`. This mirrors code from oauthlib near exactly, with the thought that I would like to break it out and submit it as a patch there.